core/sessions/ldapauth/ldap.go

```diff
@@ -0,0 +1,861 @@
+
+MFA is supported via the remote LDAP server implementation. Sufficient request time out should accomodate
+for a blocking auth call while the user responds to a potential push notification callback.
+
+*/
+
+import (
+	"crypto/subtle"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/auth"
+	"github.com/smartcontractkit/chainlink/v2/core/bridges"
+	"github.com/smartcontractkit/chainlink/v2/core/config"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/logger/audit"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/sessions"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/mathutil"
+)
+
+const (
+	LDAPUniqueMemberAttribute = "uniqueMember"
+)
+
+// implements sessions.AuthenticationProvider interface
+type ldapAuthenticator struct {
+	q           pg.Q
+	config      config.LDAP
+	lggr        logger.Logger
+	auditLogger audit.AuditLogger
+}
+
+var _ sessions.AuthenticationProvider = (*ldapAuthenticator)(nil)
+
+func NewLDAPAuthenticator(
+	db *sqlx.DB,
+	pgCfg pg.QConfig,
+	ldapCfg config.LDAP,
+	dev bool,
+	lggr logger.Logger,
+	auditLogger audit.AuditLogger,
+) (sessions.AuthenticationProvider, error) {
+	namedLogger := lggr.Named("LDAPAuthenticationProvider")
+
+	// If not chainlink dev and not tls, error
+	if !dev && !ldapCfg.ServerTLS() {
+		return nil, errors.New("LDAP Authentication driver requires TLS when running in Production mode")
+	}
+
+	// Ensure all RBAC role mappings to LDAP Groups are defined, and required fields populated, or error on startup
+	if ldapCfg.AdminUserGroupCN() == "" || ldapCfg.EditUserGroupCN() == "" ||
+		ldapCfg.RunUserGroupCN() == "" || ldapCfg.ReadUserGroupCN() == "" {
+		return nil, errors.New("LDAP Group mapping from server group name for all local RBAC role required. Set group names for `_UserGroupCN` fields")
+	}
+	if ldapCfg.ServerAddress() == "" {
+		return nil, errors.New("LDAP ServerAddress config required")
+	}
+	if ldapCfg.ReadOnlyUserLogin() == "" {
+		return nil, errors.New("LDAP ReadOnlyUserLogin config required")
+	}
+
+	ldapAuth := ldapAuthenticator{
+		q:           pg.NewQ(db, namedLogger, pgCfg),
+		config:      ldapCfg,
+		lggr:        lggr.Named("LDAPAuthenticationProvider"),
+		auditLogger: auditLogger,
+	}
+
+	// Single override of library defined global
+	ldap.DefaultTimeout = ldapCfg.QueryTimeout()
+
+	// Test initial connection and credentials
+	lggr.Infof("Attempting initial connection to configured LDAP server with bind as API user")
+	conn, err := ldapAuth.dialAndConnect()
+	if err != nil {
+		return nil, errors.Errorf("Unable to establish connection to LDAP server with provided URL and credentials: %v", err)
```

Comment:
We don't do a great job of following this, but errors typical start lower-case:
```suggestion
		return nil, errors.Errorf("unable to establish connection to LDAP server with provided URL and credentials: %v", err)
```
(with exceptions for acronyms, like `LDAP`)

Reply:
Ah thats always counter intuitive for me lol. Updated all instances in ldap and sync source files